File: android/app/src/main/kotlin/com/aiworkdeck/mobile/model/LibraryGrouping.kt
package com.aiworkdeck.mobile.model

import com.aiworkdeck.contract.ContractStates
import com.aiworkdeck.mobile.design.tr
import java.time.LocalDate
import java.time.ZoneId

data class DaySection(val day: LocalDate, val title: String, val items: List<CaptureItem>)
data class ProjectChoice(val id: String, val name: String)

object LibraryGrouping {
    fun projectId(deviceId: String, key: String) = "$deviceId:$key"

    /** 删除确认等级：按所选里最坏的桶说话。n = 该桶件数；landed 时 n = 总数。 */
    fun deleteWarningLevel(states: List<TransferState>): Pair<String, Int> {
        for (phase in ContractStates.deleteWarnOrder) {
            if (phase == "landed") break
            val n = states.count { it.phase.raw == phase }
            if (n > 0) return ContractStates.deleteWarnLevel.getValue(phase) to n
        }
        return "landed" to states.size
    }

    fun deleteWarning(items: List<CaptureItem>): String {
        val (level, n) = deleteWarningLevel(items.map { it.state })
        return tr(ContractStates.deleteWarnKey.getValue(level), mapOf("n" to n))
    }

    /**
     * 正在看的那个项目里的件。没记项目的件归到 `"unknown"` 桶——它们确实存在，
     * 只是不知道归谁，藏起来比放错地方更糟。
     */
    fun itemsIn(items: List<CaptureItem>, projectId: String?): List<CaptureItem> =
        items.filter { (it.project?.id ?: UNKNOWN_PROJECT) == projectId }

    const val UNKNOWN_PROJECT = "unknown"

    /** 按本地自然日分段，新的在前；段内按时间倒序。段头「M月d日 · N 件」。 */
    fun groupByDay(items: List<CaptureItem>, zone: ZoneId = ZoneId.systemDefault()): List<DaySection> =
        items.groupBy { it.capturedAt.atZone(zone).toLocalDate() }.entries
            .sortedByDescending { it.key }
            .map { (day, list) ->
                val sorted = list.sortedByDescending { it.capturedAt }
                DaySection(day, tr("library.dayTitle", mapOf("m" to day.monthValue, "d" to day.dayOfMonth, "n" to sorted.size)), sorted)
            }

    /** 可切换的项目：当前项目永远第一，其余有记录的按名称排；无项目的记录归「未知项目」。 */
    fun projectsIn(items: List<CaptureItem>, current: RelayProject?): List<ProjectChoice> {
        val seen = LinkedHashMap<String, String>()
        for (it in items) {
            val p = it.project
            if (p == null) seen.putIfAbsent(UNKNOWN_PROJECT, tr("library.unknownProject")) else seen.putIfAbsent(p.id, p.name)
        }
        current?.let { seen.remove(it.id) }
        val rest = seen.entries.map { ProjectChoice(it.key, it.value) }.sortedBy { it.name }
        return listOfNotNull(current?.let { ProjectChoice(it.id, it.name) }) + rest
    }
}

File: ios/Sources/Model/LibraryGrouping.swift
import Foundation

/// 图集的分组与文案，全是纯函数——界面只负责画。
enum LibraryGrouping {
    /// 可切换的项目：当前项目永远第一（哪怕还没拍），其余有记录的按名称排，「未知项目」有记录时垫底。
    static func projects(in items: [CaptureItem], current: RelayProject?) -> [LibraryProject] {
        var seen: [String: LibraryProject] = [:]
        for i in items {
            if let p = i.project { seen[p.id] = LibraryProject(p) }
            else { seen[LibraryProject.unknownID] = .unknown }
        }
        var out: [LibraryProject] = []
        if let c = current {
            out.append(LibraryProject(c))
            seen[c.id] = nil
        }
        let unknown = seen.removeValue(forKey: LibraryProject.unknownID)
        out += seen.values.sorted { $0.name.localizedStandardCompare($1.name) == .orderedAscending }
        if let unknown { out.append(unknown) }
        return out
    }

    static func items(_ items: [CaptureItem], in projectID: String) -> [CaptureItem] {
        items.filter { $0.projectID == projectID }
    }

    struct DaySection: Identifiable {
        let id: Date
        let title: String
        let items: [CaptureItem]
    }

    /// 按自然日分段，新的在前；段内按拍摄时间倒序。
    static func days(_ items: [CaptureItem], calendar: Calendar = .current) -> [DaySection] {
        let grouped = Dictionary(grouping: items) { calendar.startOfDay(for: $0.capturedAt) }
        let f = DateFormatter()
        f.calendar = calendar
        f.timeZone = calendar.timeZone
        f.dateFormat = "M月d日"
        return grouped.keys.sorted(by: >).map { day in
            let list = grouped[day]!.sorted { $0.capturedAt > $1.capturedAt }
            return DaySection(id: day, title: "\(f.string(from: day)) · \(list.count) 件", items: list)
        }
    }

    /// 删除确认等级：按所选里最坏的桶说话。n = 该桶件数；landed 时 n = 总数。
    static func deleteWarningLevel(_ states: [TransferState]) -> (level: String, n: Int) {
        for phase in ContractStates.deleteWarnOrder where phase != "landed" {
            let n = states.filter { $0.phase.rawValue == phase }.count
            if n > 0 { return (ContractStates.deleteWarnLevel[phase]!, n) }
        }
        return ("landed", states.count)
    }

    static func deleteWarning(for items: [CaptureItem]) -> String {
        let (level, n) = deleteWarningLevel(items.map(\.state))
        return tr(ContractStates.deleteWarnKey[level]!, ["n": String(n)])
    }
}
